### ISSUE TITLE
[v10] allowLogin must be true for SSO to work on signup

### ISSUE THREAD

**Issue by dshorowitz** [auth0/lock#481]
If you configure the widget with allowLogin equal to false, users from domains that should be logging in via an enterprise connection are instead allowed to establish a username/password combination. This shouldn't be allowed, even if allowLogin is false.

Expected functionality:
If the SSO domain is somedomain.com, when the user types in "email@somedomain.com" into the signup widget, the password field disappears and the user is told to sign in via the configured connection (even if allowLogin is false).

**Comment by gnandretta** [comment]
Hi, the feature you describe seems useful but if it would be confusing to allow users to log in if you  set `allowLogin` to `false`.

**Comment by dshorowitz** [comment]
Actually, without this feature the lock widget is somewhat broken for SSO users. Suppose you have a sign up lock widget on your site (that is intended ONLY to allow for sign ups, not logins) and a login lock widget on your site (that is intended ONLY to allow for logins, not sign ups).

Now suppose you have a relationship with "Enterprise X" such that all "Enterprise X" users must be managed by their Identity Provider, not by username/password.

Suppose someone from "Enterprise X" comes to your site for the first time. This person does not yet have a profile created, and has never accessed the site before. Should the person be expected to _know_ whether he/she should click on "Sign Up" or "Log In"? I don't think so.

The way the lock widget is intended to work is that if you've configured all "enterprisex.com" emails to use SSO via the auth0 dashboard, the widget automatically prevents a password from being entered and instead redirects logins to the SSO Identity Provider. But if the user clicks on "sign up" button that loads a lock widget configured with allowLogin equal to false, the "enterprisex.com" user will be allowed to create a new profile using username/password (irrespective of any settings in the dashboard)!

That's not good :) Users from domains with enterprise connections configured in the dashboard should **never** be allowed to signup with usernames and passwords. It shouldn't matter what settings you put into the lock config object.

**Comment by gnandretta** [comment]
You can set the `initialScreen` option to `"signUp"` instead of setting the `allowLogin` to `false`.

**Comment by dshorowitz** [comment]
Yes, but my point is that there should be no way for the user to customize the lock widget that allows a user to sign up with a username/password when he should be limited to logging in with an enterprise connection.

**Comment by gnandretta** [comment]
That is not a goal of the Lock API. To achieve that we'll have to remove the `allowedConnections` option.

**Comment by dshorowitz** [comment]
Sorry, I still disagree. In your own documentation (https://auth0.com/docs/saml-configuration) you say:

> A popular option is specifying e-mail domains as part of the IDP configuration. For example: adding the email domain "companyx.com" to the IDP configuration for company X will result **in all users with that email domain being routed to that IDP.**

(emphasis mine)

But in the current implementation, the lock widget does not ensure that all users from an email domain are routed to the IDP, because of the issue I identified in my original post.

**Comment by gnandretta** [comment]
That is the behavior that Lock provides out of the box. In your original post you are explicitly overriding it by disabling logins with the `allowLogin: false` option.

**Comment by dshorowitz** [comment]
Yes, I couldn't agree more :)

In my use case, I want the lock widget to be in sign up mode only (no logins), yet I still want to ensure that new sign ups are routed to the appropriate IDP. The lock widget should enforce that. If you have an email address associated with an enterprise IDP, you shouldn't be allowed to create a email/pass login in sign up mode, don't you think?

**Comment by gnandretta** [comment]
Let me rephrase my first answer (: I understand your use case and I think it is valid. The thing is that you are not really signing up when you enter an email with a domain associated to an enterprise IDP, you are logging in. And allowing logins, when you explicitly set `allowLogin: false`, might be unexpected for some users. Maybe we can add another option, or maybe we can document that in a way that doesn't leave room for confusion. In other words, this is not a feature that can be added lightly. Does that make sense?

**Comment by dshorowitz** [comment]
Makes sense. It might unexpected for some users to allow logins when allowLogin is set to false. On the other hand, it also might be unexpected for some users to see new profiles with user/pass identities when they thought they had configured the lock widget to prevent that :)

At a minimum, I'd document this very carefully!

In the interim, how can I handle my use case? Is there some event triggered in lock v10 that I can intercept so that if a user tries to sign up with an enterprise email, I can handle it appropriately client side?

**Comment by dshorowitz** [comment]
Any updates on this? I still think the lock widget should prevent username/pass signups if the email address being entered is associated with an enterprise IDP. Having the lock widget set to signUpMode only is kind of a broken feature otherwise :)

**Comment by hzalaz** [comment]
Sorry, we lost the track of this old issue. We are still sure that this is the correct behaviour for signup only and is not a broken feature since it only allows signup which is what its supposed to do. An enhancement could be that if an email is matched with an enterprise domain if blocks the user from signup since thats the only thing the widget allows in that config mode.

Another workaround is not disable login but show lock with signup as the initial screen so if the user can login with an enterprise account is able to do it.

For now will add the enhancement to our backlog to review in the future.

Sorry again for the long delay
